{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bz2\n",
    "import pickle\n",
    "from metaphlan.metaphlan import map2bbh\n",
    "import numpy as np\n",
    "\n",
    "with bz2.BZ2File( '/ssd1/wy/workspace/MetaPhlAn/test_data/databases/mpa_vJan21_TOY_CHOCOPhlAnSGB_202103.pkl', 'r' ) as a:\n",
    "    mpa_pkl = pickle.load( a )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "SGB_ANALYSIS = True\n",
    "tax_units = \"kpcofgst\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def plain_read_and_split(ofn):\n",
    "    return (l.strip().split('\\t') for l in ofn)\n",
    "reads2markers = {}\n",
    "n_metagenome_reads = None\n",
    "avg_read_length = 1 #Set to 1 if it is not calculated from read_fastx\n",
    "for r, c in plain_read_and_split(open(\"/ssd1/wy/workspace/MetaPhlAn/test_data/01.fasta.bowtie2out.txt\")):\n",
    "    if r.startswith('#') and 'nreads' in r:\n",
    "        n_metagenome_reads = int(c)\n",
    "    if r.startswith('#') and 'avg_read_length' in r:\n",
    "        avg_read_length = float(c)\n",
    "    else:\n",
    "        reads2markers[r] = c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "93.71439067912249"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "avg_read_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "class TaxClade:\n",
    "    min_cu_len = -1\n",
    "    markers2lens = None\n",
    "    stat = None\n",
    "    perc_nonzero = None\n",
    "    quantile = None\n",
    "    avoid_disqm = False\n",
    "    avg_read_length = 1\n",
    "\n",
    "    def __init__( self, name, tax_id, uncl = False):\n",
    "        self.children, self.markers2nreads = {}, {}\n",
    "        self.name, self.father = name, None\n",
    "        self.uncl, self.subcl_uncl = uncl, False\n",
    "        self.abundance, self.uncl_abundance = None, 0\n",
    "        self.nreads, self.uncl_nreads = 0, 0\n",
    "        self.tax_id = tax_id\n",
    "\n",
    "    def add_child( self, name, tax_id ):\n",
    "        new_clade = TaxClade( name, tax_id )\n",
    "        self.children[name] = new_clade\n",
    "        new_clade.father = self\n",
    "        return new_clade\n",
    "\n",
    "\n",
    "    def get_terminals( self ):\n",
    "        terms = []\n",
    "        if not self.children:\n",
    "            return [self]\n",
    "        for c in self.children.values():\n",
    "            terms += c.get_terminals()\n",
    "        return terms\n",
    "\n",
    "    def get_full_taxids( self ):\n",
    "        fullname = ['']\n",
    "        if self.tax_id:\n",
    "            fullname = [self.tax_id]\n",
    "        cl = self.father\n",
    "        while cl:\n",
    "            fullname = [cl.tax_id] + fullname\n",
    "            cl = cl.father\n",
    "        return \"|\".join(fullname[1:])\n",
    "\n",
    "    def get_full_name( self ):\n",
    "        fullname = [self.name]\n",
    "        cl = self.father\n",
    "        while cl:\n",
    "            fullname = [cl.name] + fullname\n",
    "            cl = cl.father\n",
    "        return \"|\".join(fullname[1:])\n",
    "\n",
    "    def get_normalized_counts( self ):\n",
    "        return [(m,float(n)*1000.0/(np.absolute(self.markers2lens[m] - self.avg_read_length) +1) )\n",
    "                    for m,n in self.markers2nreads.items()]\n",
    "\n",
    "    def compute_mapped_reads( self ):    \n",
    "        tax_level = 't__' if SGB_ANALYSIS else 's__'\n",
    "        if self.nreads != 0 or self.name.startswith(tax_level):\n",
    "            return self.nreads\n",
    "        for c in self.children.values():\n",
    "            self.nreads += c.compute_mapped_reads()\n",
    "        return self.nreads\n",
    "        \n",
    "    def compute_abundance( self ):\n",
    "        if self.abundance is not None: return self.abundance\n",
    "\n",
    "        sum_ab = sum([c.compute_abundance() for c in self.children.values()])\n",
    "\n",
    "        # rat_nreads = sorted([(self.markers2lens[marker], n_reads)\n",
    "        #                             for marker,n_reads in self.markers2nreads.items()],\n",
    "        #                                     key = lambda x: x[1])\n",
    "\n",
    "        rat_nreads, removed = [], []\n",
    "        for marker, n_reads in sorted(self.markers2nreads.items(),key=lambda x:x[0]):\n",
    "            misidentified = False\n",
    "\n",
    "            if not self.avoid_disqm:\n",
    "                for ext in self.markers2exts[marker]:\n",
    "                    ext_clade = self.taxa2clades[ext]\n",
    "                    m2nr = ext_clade.markers2nreads\n",
    "                    \n",
    "                    tocladetmp = ext_clade\n",
    "                    while len(tocladetmp.children) == 1:\n",
    "                        tocladetmp = list(tocladetmp.children.values())[0]\n",
    "                        m2nr = tocladetmp.markers2nreads\n",
    "\n",
    "                    nonzeros = sum([v>0 for v in m2nr.values()])\n",
    "                    if len(m2nr):\n",
    "                        if float(nonzeros) / len(m2nr) > self.perc_nonzero:\n",
    "                            misidentified = True\n",
    "                            removed.append( (self.markers2lens[marker],n_reads) )\n",
    "                            break\n",
    "            if not misidentified:\n",
    "                rat_nreads.append( (self.markers2lens[marker],n_reads) )\n",
    "\n",
    "        if not self.avoid_disqm and len(removed):\n",
    "            n_rat_nreads = float(len(rat_nreads))\n",
    "            n_removed = float(len(removed))\n",
    "            n_tot = n_rat_nreads + n_removed\n",
    "            n_ripr = 10\n",
    "\n",
    "            if len(self.get_terminals()) < 2:\n",
    "                n_ripr = 0\n",
    "\n",
    "            if \"k__Viruses\" in self.get_full_name():\n",
    "                n_ripr = 0\n",
    "\n",
    "            if n_rat_nreads < n_ripr and n_tot > n_rat_nreads:\n",
    "                rat_nreads += removed[:n_ripr-int(n_rat_nreads)]\n",
    "\n",
    "\n",
    "        rat_nreads = sorted(rat_nreads, key = lambda x: x[1])\n",
    "\n",
    "        rat_v,nreads_v = zip(*rat_nreads) if rat_nreads else ([],[])\n",
    "        rat, nrawreads, loc_ab = float(sum(rat_v)) or -1.0, sum(nreads_v), 0.0\n",
    "        quant = int(self.quantile*len(rat_nreads))\n",
    "        ql,qr,qn = (quant,-quant,quant) if quant else (None,None,0)\n",
    "\n",
    "        if not SGB_ANALYSIS and self.name[0] == 't' and (len(self.father.children) > 1 or \"_sp\" in self.father.name or \"k__Viruses\" in self.get_full_name()):\n",
    "            non_zeros = float(len([n for r,n in rat_nreads if n > 0]))\n",
    "            nreads = float(len(rat_nreads))\n",
    "            if nreads == 0.0 or non_zeros / nreads < 0.7:\n",
    "                self.abundance = 0.0\n",
    "                return 0.0\n",
    "\n",
    "        if rat < 0.0:\n",
    "            pass\n",
    "        elif self.stat == 'avg_g' or (not qn and self.stat in ['wavg_g','tavg_g']):\n",
    "            loc_ab = nrawreads / rat if rat >= 0 else 0.0\n",
    "        elif self.stat == 'avg_l' or (not qn and self.stat in ['wavg_l','tavg_l']):\n",
    "            loc_ab = np.mean([float(n)/(np.absolute(r - self.avg_read_length) + 1) for r,n in rat_nreads])\n",
    "        elif self.stat == 'tavg_g':\n",
    "            wnreads = sorted([(float(n)/(np.absolute(r-self.avg_read_length)+1),(np.absolute(r - self.avg_read_length)+1) ,n) for r,n in rat_nreads], key=lambda x:x[0])\n",
    "            den,num = zip(*[v[1:] for v in wnreads[ql:qr]])\n",
    "            loc_ab = float(sum(num))/float(sum(den)) if any(den) else 0.0\n",
    "        elif self.stat == 'tavg_l':\n",
    "            loc_ab = np.mean(sorted([float(n)/(np.absolute(r - self.avg_read_length) + 1) for r,n in rat_nreads])[ql:qr])\n",
    "        elif self.stat == 'wavg_g':\n",
    "            vmin, vmax = nreads_v[ql], nreads_v[qr]\n",
    "            wnreads = [vmin]*qn+list(nreads_v[ql:qr])+[vmax]*qn\n",
    "            loc_ab = float(sum(wnreads)) / rat\n",
    "        elif self.stat == 'wavg_l':\n",
    "            wnreads = sorted([float(n)/(np.absolute(r - self.avg_read_length) + 1) for r,n in rat_nreads])\n",
    "            vmin, vmax = wnreads[ql], wnreads[qr]\n",
    "            wnreads = [vmin]*qn+list(wnreads[ql:qr])+[vmax]*qn\n",
    "            loc_ab = np.mean(wnreads)\n",
    "        elif self.stat == 'med':\n",
    "            loc_ab = np.median(sorted([float(n)/(np.absolute(r - self.avg_read_length) +1) for r,n in rat_nreads])[ql:qr])\n",
    "\n",
    "        self.abundance = loc_ab\n",
    "        if rat < self.min_cu_len and self.children:\n",
    "            self.abundance = sum_ab\n",
    "        elif loc_ab < sum_ab:\n",
    "            self.abundance = sum_ab\n",
    "\n",
    "        if self.abundance > sum_ab and self.children: # *1.1??\n",
    "            self.uncl_abundance = self.abundance - sum_ab\n",
    "        self.subcl_uncl = not self.children and self.name[0] not in tax_units[-2:]\n",
    "\n",
    "        return self.abundance\n",
    "\n",
    "    def get_all_abundances( self ):\n",
    "        ret = [(self.name, self.tax_id, self.abundance)]\n",
    "        if self.uncl_abundance > 0.0:\n",
    "            lchild = list(self.children.values())[0].name[:3]\n",
    "            ret += [(lchild+self.name[3:]+\"_unclassified\", \"\", self.uncl_abundance)]\n",
    "        if self.subcl_uncl and self.name[0] != tax_units[-2]:\n",
    "            cind = tax_units.index( self.name[0] )\n",
    "            ret += [(   tax_units[cind+1]+self.name[1:]+\"_unclassified\",\"\",\n",
    "                        self.abundance)]\n",
    "        for c in self.children.values():\n",
    "            ret += c.get_all_abundances()\n",
    "        return ret\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "class TaxTree:\n",
    "    def __init__( self, mpa, markers_to_ignore = None ): #, min_cu_len ):\n",
    "        self.root = TaxClade( \"root\", 0)\n",
    "        self.all_clades, self.markers2lens, self.markers2clades, self.taxa2clades, self.markers2exts = {}, {}, {}, {}, {}\n",
    "        TaxClade.markers2lens = self.markers2lens\n",
    "        TaxClade.markers2exts = self.markers2exts\n",
    "        TaxClade.taxa2clades = self.taxa2clades\n",
    "        self.avg_read_length = 1\n",
    "\n",
    "        for clade, value in mpa['taxonomy'].items():\n",
    "            clade = clade.strip().split(\"|\")\n",
    "            if isinstance(value,tuple):\n",
    "                taxids, lenc = value\n",
    "                taxids = taxids.strip().split(\"|\")\n",
    "            if isinstance(value,int):\n",
    "                lenc = value\n",
    "                taxids = None\n",
    "\n",
    "            father = self.root\n",
    "            for i in range(len(clade)):\n",
    "                clade_lev = clade[i]\n",
    "                if SGB_ANALYSIS:\n",
    "                    clade_taxid = taxids[i] if i < 8 and taxids is not None else None\n",
    "                else:\n",
    "                    clade_taxid = taxids[i] if i < 7 and taxids is not None else None\n",
    "                if not clade_lev in father.children:\n",
    "                    father.add_child(clade_lev, tax_id=clade_taxid)\n",
    "                    self.all_clades[clade_lev] = father.children[clade_lev]\n",
    "                if SGB_ANALYSIS: father = father.children[clade_lev]\n",
    "                if clade_lev[0] == \"t\":\n",
    "                    self.taxa2clades[clade_lev[3:]] = father\n",
    "                if not SGB_ANALYSIS: father = father.children[clade_lev]\n",
    "                if clade_lev[0] == \"t\":\n",
    "                    father.glen = lenc\n",
    "\n",
    "        def add_lens( node ):\n",
    "            if not node.children:\n",
    "                return node.glen\n",
    "            lens = []\n",
    "            for c in node.children.values():\n",
    "                lens.append( add_lens( c ) )\n",
    "            node.glen = min(np.mean(lens), np.median(lens))\n",
    "            return node.glen\n",
    "        \n",
    "        add_lens(self.root)\n",
    "\n",
    "        # for k,p in mpa_pkl['markers'].items():\n",
    "        for k, p in mpa['markers'].items():\n",
    "            if k in markers_to_ignore:\n",
    "                continue\n",
    "            self.markers2lens[k] = p['len']\n",
    "            self.markers2clades[k] = p['clade']\n",
    "            self.add_reads(k, 0)\n",
    "            self.markers2exts[k] = p['ext']\n",
    "\n",
    "    def set_min_cu_len( self, min_cu_len ):\n",
    "        TaxClade.min_cu_len = min_cu_len\n",
    "\n",
    "    def set_stat( self, stat, quantile, perc_nonzero, avg_read_length, avoid_disqm = False):\n",
    "        TaxClade.stat = stat\n",
    "        TaxClade.perc_nonzero = perc_nonzero\n",
    "        TaxClade.quantile = quantile\n",
    "        TaxClade.avoid_disqm = avoid_disqm\n",
    "        TaxClade.avg_read_length = avg_read_length\n",
    "\n",
    "    def add_reads(  self, marker, n,\n",
    "                    add_viruses = False,\n",
    "                    ignore_eukaryotes = False,\n",
    "                    ignore_bacteria = False, ignore_archaea = False, \n",
    "                    ignore_ksgbs = False, ignore_usgbs = False  ):\n",
    "        clade = self.markers2clades[marker]\n",
    "        cl = self.all_clades[clade]\n",
    "        if ignore_bacteria or ignore_archaea or ignore_eukaryotes:\n",
    "            cn = cl.get_full_name()\n",
    "            if ignore_archaea and cn.startswith(\"k__Archaea\"):\n",
    "                return (None, None)\n",
    "            if ignore_bacteria and cn.startswith(\"k__Bacteria\"):\n",
    "                return (None, None)\n",
    "            if ignore_eukaryotes and cn.startswith(\"k__Eukaryota\"):\n",
    "                return (None, None)\n",
    "        if not SGB_ANALYSIS and not add_viruses:\n",
    "            cn = cl.get_full_name()\n",
    "            if not add_viruses and cn.startswith(\"k__Vir\"):\n",
    "                return (None, None)\n",
    "        if SGB_ANALYSIS and (ignore_ksgbs or ignore_usgbs):\n",
    "            cn = cl.get_full_name()\n",
    "            if ignore_ksgbs and not '_SGB' in cn.split('|')[-2]:\n",
    "                return (None, None)\n",
    "            if ignore_usgbs and '_SGB' in cn.split('|')[-2]:\n",
    "                return (None, None)\n",
    "        # while len(cl.children) == 1:\n",
    "            # cl = list(cl.children.values())[0]\n",
    "        cl.markers2nreads[marker] = n\n",
    "        return (cl.get_full_name(), cl.get_full_taxids(), )\n",
    "\n",
    "\n",
    "    def markers2counts( self ):\n",
    "        m2c = {}\n",
    "        for _ ,v in self.all_clades.items():\n",
    "            for m,c in v.markers2nreads.items():\n",
    "                m2c[m] = c\n",
    "        return m2c\n",
    "\n",
    "    def clade_profiles( self, tax_lev, get_all = False  ):\n",
    "        cl2pr = {}\n",
    "        for k,v in self.all_clades.items():\n",
    "            if tax_lev and not k.startswith(tax_lev):\n",
    "                continue\n",
    "            prof = v.get_normalized_counts()\n",
    "            if not get_all and ( len(prof) < 1 or not sum([p[1] for p in prof]) > 0.0 ):\n",
    "                continue\n",
    "            cl2pr[v.get_full_name()] = prof\n",
    "        return cl2pr\n",
    "\n",
    "    def relative_abundances( self, tax_lev  ):\n",
    "        clade2abundance_n = dict([(tax_label, clade) for tax_label, clade in self.all_clades.items()\n",
    "                    if tax_label.startswith(\"k__\") and not clade.uncl])\n",
    "\n",
    "        clade2abundance, clade2est_nreads, tot_ab, tot_reads = {}, {}, 0.0, 0\n",
    "\n",
    "        for tax_label, clade in clade2abundance_n.items():\n",
    "            tot_ab += clade.compute_abundance()\n",
    "\n",
    "        for tax_label, clade in clade2abundance_n.items():\n",
    "            for clade_label, tax_id, abundance in sorted(clade.get_all_abundances(), key=lambda pars:pars[0]):\n",
    "                if SGB_ANALYSIS or clade_label[:3] != 't__':\n",
    "                    if not tax_lev:\n",
    "                        if clade_label not in self.all_clades:\n",
    "                            to = tax_units.index(clade_label[0])\n",
    "                            t = tax_units[to-1]\n",
    "                            clade_label = t + clade_label.split(\"_unclassified\")[0][1:]\n",
    "                            tax_id = self.all_clades[clade_label].get_full_taxids()\n",
    "                            clade_label = self.all_clades[clade_label].get_full_name()\n",
    "                            spl = clade_label.split(\"|\")\n",
    "                            clade_label = \"|\".join(spl+[tax_units[to]+spl[-1][1:]+\"_unclassified\"])\n",
    "                            glen = self.all_clades[spl[-1]].glen\n",
    "                        else:\n",
    "                            glen = self.all_clades[clade_label].glen\n",
    "                            tax_id = self.all_clades[clade_label].get_full_taxids()\n",
    "                            tax_level = 't__' if SGB_ANALYSIS else 's__' \n",
    "                            if tax_level in clade_label and abundance > 0:\n",
    "                                self.all_clades[clade_label].nreads = int(np.floor(abundance*glen))\n",
    "\n",
    "                            clade_label = self.all_clades[clade_label].get_full_name()\n",
    "                    elif not clade_label.startswith(tax_lev):\n",
    "                        if clade_label in self.all_clades:\n",
    "                            glen = self.all_clades[clade_label].glen\n",
    "                        else:\n",
    "                            glen = 1.0\n",
    "                        continue\n",
    "                    clade2abundance[(clade_label, tax_id)] = abundance\n",
    "        \n",
    "        for tax_label, clade in clade2abundance_n.items():\n",
    "            tot_reads += clade.compute_mapped_reads()\n",
    "\n",
    "        for clade_label, clade in self.all_clades.items():\n",
    "            if SGB_ANALYSIS or clade.name[:3] != 't__':\n",
    "                nreads = clade.nreads\n",
    "                clade_label = clade.get_full_name()\n",
    "                tax_id = clade.get_full_taxids()\n",
    "                clade2est_nreads[(clade_label, tax_id)] = nreads\n",
    "\n",
    "        ret_d = dict([( tax, float(abundance) / tot_ab if tot_ab else 0.0) for tax, abundance in clade2abundance.items()])\n",
    "\n",
    "        ret_r = dict([( tax, (abundance, clade2est_nreads[tax] )) for tax, abundance in clade2abundance.items() if tax in clade2est_nreads])\n",
    "\n",
    "        if tax_lev:\n",
    "            ret_d[(\"UNCLASSIFIED\", '-1')] = 1.0 - sum(ret_d.values())\n",
    "        return ret_d, ret_r, tot_reads\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: The number of reads in the sample (None) is below the recommended minimum of 10,000 reads."
     ]
    }
   ],
   "source": [
    "# REPORT_MERGED = mpa_pkl.get('merged_taxon',False)\n",
    "tree = TaxTree( mpa_pkl, {} )\n",
    "tree.set_min_cu_len( 2000 )\n",
    "# minimum total nucleotide length for the markers in a clade for estimating the abundance without considering sub-clade abundances\n",
    "markers2reads, n_metagenome_reads, avg_read_length = map2bbh(\"/ssd1/wy/workspace/MetaPhlAn/test_data/01.fasta.bowtie2out.txt\", \n",
    "                                                           5, \"bowtie2out\", None, None, None, '1992')\n",
    "tree.set_stat( 'tavg_g',0.2, 0.33, avg_read_length,False)\n",
    "\n",
    "map_out = []\n",
    "for marker,reads in sorted(markers2reads.items(), key=lambda pars: pars[0]):\n",
    "    if marker not in tree.markers2lens:\n",
    "        continue\n",
    "    tax_seq, ids_seq = tree.add_reads( marker, len(reads),\n",
    "                                add_viruses = False,\n",
    "                                ignore_eukaryotes = False,\n",
    "                                ignore_bacteria = False,\n",
    "                                ignore_archaea = False,\n",
    "                                ignore_ksgbs = False,\n",
    "                                ignore_usgbs = False\n",
    "                                )\n",
    "    if tax_seq:\n",
    "        map_out +=[\"\\t\".join([r,tax_seq, ids_seq]) for r in sorted(reads)]\n",
    "        \n",
    "cl2ab, _, tot_nreads = tree.relative_abundances(\n",
    "                        'a'+\"__\" if 'a' != 'a' else None )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria 2|201174|1760 1.0\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Corynebacteriales|f__Corynebacteriaceae 2|201174|1760|85007|1653 0.6086077978332424\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Micrococcales|f__Micrococcaceae 2|201174|1760|85006|1268 0.3913922021667576\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Corynebacteriales|f__Corynebacteriaceae|g__Corynebacterium 2|201174|1760|85007|1653|1716 0.6086077978332424\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Micrococcales|f__Micrococcaceae|g__Rothia 2|201174|1760|85006|1268|32207 0.3913922021667576\n",
      "k__Bacteria 2 1.0\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Corynebacteriales 2|201174|1760|85007 0.6086077978332424\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Micrococcales 2|201174|1760|85006 0.3913922021667576\n",
      "k__Bacteria|p__Actinobacteria 2|201174 1.0\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Corynebacteriales|f__Corynebacteriaceae|g__Corynebacterium|s__Corynebacterium_matruchotii 2|201174|1760|85007|1653|1716|43768 0.6086077978332424\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Micrococcales|f__Micrococcaceae|g__Rothia|s__Rothia_dentocariosa 2|201174|1760|85006|1268|32207|2047 0.3913922021667576\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Micrococcales|f__Micrococcaceae|g__Rothia|s__Rothia_dentocariosa|t__SGB16988_group 2|201174|1760|85006|1268|32207|2047| 0.3913922021667576\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Corynebacteriales|f__Corynebacteriaceae|g__Corynebacterium|s__Corynebacterium_matruchotii|t__SGB17007 2|201174|1760|85007|1653|1716|43768| 0.6086077978332424\n"
     ]
    }
   ],
   "source": [
    "for (taxstr, taxid), relab in cl2ab.items():\n",
    "   if relab > 0.0:\n",
    "      print(taxstr,taxid,relab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "outpred = [(taxstr, taxid,round(relab*100.0,5)) for (taxstr, taxid), relab in cl2ab.items() if relab > 0.0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k__Bacteria\t2\t100.0\t\n",
      "\n",
      "k__Bacteria|p__Actinobacteria\t2|201174\t100.0\t\n",
      "\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria\t2|201174|1760\t100.0\t\n",
      "\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Corynebacteriales\t2|201174|1760|85007\t60.86078\t\n",
      "\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Micrococcales\t2|201174|1760|85006\t39.13922\t\n",
      "\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Corynebacteriales|f__Corynebacteriaceae\t2|201174|1760|85007|1653\t60.86078\t\n",
      "\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Micrococcales|f__Micrococcaceae\t2|201174|1760|85006|1268\t39.13922\t\n",
      "\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Corynebacteriales|f__Corynebacteriaceae|g__Corynebacterium\t2|201174|1760|85007|1653|1716\t60.86078\t\n",
      "\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Micrococcales|f__Micrococcaceae|g__Rothia\t2|201174|1760|85006|1268|32207\t39.13922\t\n",
      "\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Corynebacteriales|f__Corynebacteriaceae|g__Corynebacterium|s__Corynebacterium_matruchotii\t2|201174|1760|85007|1653|1716|43768\t60.86078\t\n",
      "\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Micrococcales|f__Micrococcaceae|g__Rothia|s__Rothia_dentocariosa\t2|201174|1760|85006|1268|32207|2047\t39.13922\t\n",
      "\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Corynebacteriales|f__Corynebacteriaceae|g__Corynebacterium|s__Corynebacterium_matruchotii|t__SGB17007\t2|201174|1760|85007|1653|1716|43768|\t60.86078\t\n",
      "\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Micrococcales|f__Micrococcaceae|g__Rothia|s__Rothia_dentocariosa|t__SGB16988_group\t2|201174|1760|85006|1268|32207|2047|\t39.13922\t\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for clade, taxid, relab in sorted(  outpred, reverse=True,\n",
    "                                        key=lambda x:x[2]+(100.0*(8-(x[0].count(\"|\"))))):\n",
    "    add_repr = ''\n",
    "    print( \"\\t\".join( [clade, \n",
    "                                taxid, \n",
    "                                str(relab*1.0), \n",
    "                                add_repr\n",
    "                            ] ) + \"\\n\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria 2|201174|1760 1.0\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Corynebacteriales|f__Corynebacteriaceae 2|201174|1760|85007|1653 0.6086077978332424\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Micrococcales|f__Micrococcaceae 2|201174|1760|85006|1268 0.3913922021667576\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Corynebacteriales|f__Corynebacteriaceae|g__Corynebacterium 2|201174|1760|85007|1653|1716 0.6086077978332424\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Micrococcales|f__Micrococcaceae|g__Rothia 2|201174|1760|85006|1268|32207 0.3913922021667576\n",
      "k__Bacteria 2 1.0\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Corynebacteriales 2|201174|1760|85007 0.6086077978332424\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Micrococcales 2|201174|1760|85006 0.3913922021667576\n",
      "k__Bacteria|p__Actinobacteria 2|201174 1.0\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Corynebacteriales|f__Corynebacteriaceae|g__Corynebacterium|s__Corynebacterium_matruchotii 2|201174|1760|85007|1653|1716|43768 0.6086077978332424\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Micrococcales|f__Micrococcaceae|g__Rothia|s__Rothia_dentocariosa 2|201174|1760|85006|1268|32207|2047 0.3913922021667576\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Micrococcales|f__Micrococcaceae|g__Rothia|s__Rothia_dentocariosa|t__SGB16988_group 2|201174|1760|85006|1268|32207|2047| 0.3913922021667576\n",
      "k__Bacteria|p__Actinobacteria|c__Actinobacteria|o__Corynebacteriales|f__Corynebacteriaceae|g__Corynebacterium|s__Corynebacterium_matruchotii|t__SGB17007 2|201174|1760|85007|1653|1716|43768| 0.6086077978332424\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_stream =  open(\"a.txt\",\"w\") #sys.stdout\n",
    "MPA2_OUTPUT = False\n",
    "CAMI_OUTPUT = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unsupported operand type(s) for *: 'NoneType' and 'int'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[37], line 7\u001b[0m\n\u001b[1;32m      5\u001b[0m outf\u001b[39m.\u001b[39mwrite(\u001b[39m'\u001b[39m\u001b[39m#\u001b[39m\u001b[39m'\u001b[39m \u001b[39m+\u001b[39m \u001b[39m'\u001b[39m\u001b[39m\\t\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m.\u001b[39mjoin((\u001b[39m'\u001b[39m\u001b[39mSampleID\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mMetaphlan_Analysis\u001b[39m\u001b[39m'\u001b[39m)) \u001b[39m+\u001b[39m \u001b[39m'\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m)\n\u001b[1;32m      6\u001b[0m outf\u001b[39m.\u001b[39mwrite(\u001b[39m'\u001b[39m\u001b[39m#clade_name\u001b[39m\u001b[39m\\t\u001b[39;00m\u001b[39mNCBI_tax_id\u001b[39m\u001b[39m\\t\u001b[39;00m\u001b[39mrelative_abundance\u001b[39m\u001b[39m\\t\u001b[39;00m\u001b[39madditional_species\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m)\n\u001b[0;32m----> 7\u001b[0m cl2ab, _, tot_nreads \u001b[39m=\u001b[39m tree\u001b[39m.\u001b[39;49mrelative_abundances(\n\u001b[1;32m      8\u001b[0m                         \u001b[39m'\u001b[39;49m\u001b[39ma\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39m+\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m__\u001b[39;49m\u001b[39m\"\u001b[39;49m \u001b[39mif\u001b[39;49;00m \u001b[39m'\u001b[39;49m\u001b[39mtax_lev\u001b[39;49m\u001b[39m'\u001b[39;49m \u001b[39m!=\u001b[39;49m \u001b[39m'\u001b[39;49m\u001b[39ma\u001b[39;49m\u001b[39m'\u001b[39;49m \u001b[39melse\u001b[39;49;00m \u001b[39mNone\u001b[39;49;00m )\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:767\u001b[0m, in \u001b[0;36mTaxTree.relative_abundances\u001b[0;34m(self, tax_lev)\u001b[0m\n\u001b[1;32m    764\u001b[0m clade2abundance, clade2est_nreads, tot_ab, tot_reads \u001b[39m=\u001b[39m {}, {}, \u001b[39m0.0\u001b[39m, \u001b[39m0\u001b[39m\n\u001b[1;32m    766\u001b[0m \u001b[39mfor\u001b[39;00m tax_label, clade \u001b[39min\u001b[39;00m clade2abundance_n\u001b[39m.\u001b[39mitems():\n\u001b[0;32m--> 767\u001b[0m     tot_ab \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m clade\u001b[39m.\u001b[39;49mcompute_abundance()\n\u001b[1;32m    769\u001b[0m \u001b[39mfor\u001b[39;00m tax_label, clade \u001b[39min\u001b[39;00m clade2abundance_n\u001b[39m.\u001b[39mitems():\n\u001b[1;32m    770\u001b[0m     \u001b[39mfor\u001b[39;00m clade_label, tax_id, abundance \u001b[39min\u001b[39;00m \u001b[39msorted\u001b[39m(clade\u001b[39m.\u001b[39mget_all_abundances(), key\u001b[39m=\u001b[39m\u001b[39mlambda\u001b[39;00m pars:pars[\u001b[39m0\u001b[39m]):\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:538\u001b[0m, in \u001b[0;36mTaxClade.compute_abundance\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcompute_abundance\u001b[39m( \u001b[39mself\u001b[39m ):\n\u001b[1;32m    536\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m: \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance\n\u001b[0;32m--> 538\u001b[0m     sum_ab \u001b[39m=\u001b[39m \u001b[39msum\u001b[39m([c\u001b[39m.\u001b[39mcompute_abundance() \u001b[39mfor\u001b[39;00m c \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchildren\u001b[39m.\u001b[39mvalues()])\n\u001b[1;32m    540\u001b[0m     \u001b[39m# rat_nreads = sorted([(self.markers2lens[marker], n_reads)\u001b[39;00m\n\u001b[1;32m    541\u001b[0m     \u001b[39m#                             for marker,n_reads in self.markers2nreads.items()],\u001b[39;00m\n\u001b[1;32m    542\u001b[0m     \u001b[39m#                                     key = lambda x: x[1])\u001b[39;00m\n\u001b[1;32m    544\u001b[0m     rat_nreads, removed \u001b[39m=\u001b[39m [], []\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:538\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcompute_abundance\u001b[39m( \u001b[39mself\u001b[39m ):\n\u001b[1;32m    536\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m: \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance\n\u001b[0;32m--> 538\u001b[0m     sum_ab \u001b[39m=\u001b[39m \u001b[39msum\u001b[39m([c\u001b[39m.\u001b[39;49mcompute_abundance() \u001b[39mfor\u001b[39;00m c \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchildren\u001b[39m.\u001b[39mvalues()])\n\u001b[1;32m    540\u001b[0m     \u001b[39m# rat_nreads = sorted([(self.markers2lens[marker], n_reads)\u001b[39;00m\n\u001b[1;32m    541\u001b[0m     \u001b[39m#                             for marker,n_reads in self.markers2nreads.items()],\u001b[39;00m\n\u001b[1;32m    542\u001b[0m     \u001b[39m#                                     key = lambda x: x[1])\u001b[39;00m\n\u001b[1;32m    544\u001b[0m     rat_nreads, removed \u001b[39m=\u001b[39m [], []\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:538\u001b[0m, in \u001b[0;36mTaxClade.compute_abundance\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcompute_abundance\u001b[39m( \u001b[39mself\u001b[39m ):\n\u001b[1;32m    536\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m: \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance\n\u001b[0;32m--> 538\u001b[0m     sum_ab \u001b[39m=\u001b[39m \u001b[39msum\u001b[39m([c\u001b[39m.\u001b[39mcompute_abundance() \u001b[39mfor\u001b[39;00m c \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchildren\u001b[39m.\u001b[39mvalues()])\n\u001b[1;32m    540\u001b[0m     \u001b[39m# rat_nreads = sorted([(self.markers2lens[marker], n_reads)\u001b[39;00m\n\u001b[1;32m    541\u001b[0m     \u001b[39m#                             for marker,n_reads in self.markers2nreads.items()],\u001b[39;00m\n\u001b[1;32m    542\u001b[0m     \u001b[39m#                                     key = lambda x: x[1])\u001b[39;00m\n\u001b[1;32m    544\u001b[0m     rat_nreads, removed \u001b[39m=\u001b[39m [], []\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:538\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcompute_abundance\u001b[39m( \u001b[39mself\u001b[39m ):\n\u001b[1;32m    536\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m: \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance\n\u001b[0;32m--> 538\u001b[0m     sum_ab \u001b[39m=\u001b[39m \u001b[39msum\u001b[39m([c\u001b[39m.\u001b[39;49mcompute_abundance() \u001b[39mfor\u001b[39;00m c \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchildren\u001b[39m.\u001b[39mvalues()])\n\u001b[1;32m    540\u001b[0m     \u001b[39m# rat_nreads = sorted([(self.markers2lens[marker], n_reads)\u001b[39;00m\n\u001b[1;32m    541\u001b[0m     \u001b[39m#                             for marker,n_reads in self.markers2nreads.items()],\u001b[39;00m\n\u001b[1;32m    542\u001b[0m     \u001b[39m#                                     key = lambda x: x[1])\u001b[39;00m\n\u001b[1;32m    544\u001b[0m     rat_nreads, removed \u001b[39m=\u001b[39m [], []\n",
      "    \u001b[0;31m[... skipping similar frames: <listcomp> at line 538 (4 times), TaxClade.compute_abundance at line 538 (4 times)]\u001b[0m\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:538\u001b[0m, in \u001b[0;36mTaxClade.compute_abundance\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcompute_abundance\u001b[39m( \u001b[39mself\u001b[39m ):\n\u001b[1;32m    536\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m: \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance\n\u001b[0;32m--> 538\u001b[0m     sum_ab \u001b[39m=\u001b[39m \u001b[39msum\u001b[39m([c\u001b[39m.\u001b[39mcompute_abundance() \u001b[39mfor\u001b[39;00m c \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchildren\u001b[39m.\u001b[39mvalues()])\n\u001b[1;32m    540\u001b[0m     \u001b[39m# rat_nreads = sorted([(self.markers2lens[marker], n_reads)\u001b[39;00m\n\u001b[1;32m    541\u001b[0m     \u001b[39m#                             for marker,n_reads in self.markers2nreads.items()],\u001b[39;00m\n\u001b[1;32m    542\u001b[0m     \u001b[39m#                                     key = lambda x: x[1])\u001b[39;00m\n\u001b[1;32m    544\u001b[0m     rat_nreads, removed \u001b[39m=\u001b[39m [], []\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:538\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcompute_abundance\u001b[39m( \u001b[39mself\u001b[39m ):\n\u001b[1;32m    536\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m: \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance\n\u001b[0;32m--> 538\u001b[0m     sum_ab \u001b[39m=\u001b[39m \u001b[39msum\u001b[39m([c\u001b[39m.\u001b[39;49mcompute_abundance() \u001b[39mfor\u001b[39;00m c \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchildren\u001b[39m.\u001b[39mvalues()])\n\u001b[1;32m    540\u001b[0m     \u001b[39m# rat_nreads = sorted([(self.markers2lens[marker], n_reads)\u001b[39;00m\n\u001b[1;32m    541\u001b[0m     \u001b[39m#                             for marker,n_reads in self.markers2nreads.items()],\u001b[39;00m\n\u001b[1;32m    542\u001b[0m     \u001b[39m#                                     key = lambda x: x[1])\u001b[39;00m\n\u001b[1;32m    544\u001b[0m     rat_nreads, removed \u001b[39m=\u001b[39m [], []\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:587\u001b[0m, in \u001b[0;36mTaxClade.compute_abundance\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    585\u001b[0m rat_v,nreads_v \u001b[39m=\u001b[39m \u001b[39mzip\u001b[39m(\u001b[39m*\u001b[39mrat_nreads) \u001b[39mif\u001b[39;00m rat_nreads \u001b[39melse\u001b[39;00m ([],[])\n\u001b[1;32m    586\u001b[0m rat, nrawreads, loc_ab \u001b[39m=\u001b[39m \u001b[39mfloat\u001b[39m(\u001b[39msum\u001b[39m(rat_v)) \u001b[39mor\u001b[39;00m \u001b[39m-\u001b[39m\u001b[39m1.0\u001b[39m, \u001b[39msum\u001b[39m(nreads_v), \u001b[39m0.0\u001b[39m\n\u001b[0;32m--> 587\u001b[0m quant \u001b[39m=\u001b[39m \u001b[39mint\u001b[39m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mquantile\u001b[39m*\u001b[39;49m\u001b[39mlen\u001b[39;49m(rat_nreads))\n\u001b[1;32m    588\u001b[0m ql,qr,qn \u001b[39m=\u001b[39m (quant,\u001b[39m-\u001b[39mquant,quant) \u001b[39mif\u001b[39;00m quant \u001b[39melse\u001b[39;00m (\u001b[39mNone\u001b[39;00m,\u001b[39mNone\u001b[39;00m,\u001b[39m0\u001b[39m)\n\u001b[1;32m    590\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m SGB_ANALYSIS \u001b[39mand\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mname[\u001b[39m0\u001b[39m] \u001b[39m==\u001b[39m \u001b[39m'\u001b[39m\u001b[39mt\u001b[39m\u001b[39m'\u001b[39m \u001b[39mand\u001b[39;00m (\u001b[39mlen\u001b[39m(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfather\u001b[39m.\u001b[39mchildren) \u001b[39m>\u001b[39m \u001b[39m1\u001b[39m \u001b[39mor\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39m_sp\u001b[39m\u001b[39m\"\u001b[39m \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfather\u001b[39m.\u001b[39mname \u001b[39mor\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39mk__Viruses\u001b[39m\u001b[39m\"\u001b[39m \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mget_full_name()):\n",
      "\u001b[0;31mTypeError\u001b[0m: unsupported operand type(s) for *: 'NoneType' and 'int'"
     ]
    }
   ],
   "source": [
    "with open(\"a.txt\",\"w\") as outf:\n",
    "    outf.write('#{}\\n'.format('mpa_vJan21_TOY_CHOCOPhlAnSGB_202103'))\n",
    "    outf.write('#{}\\n'.format(' '.join(sys.argv)))\n",
    "    outf.write('#{} reads processed\\n'.format(n_metagenome_reads))\n",
    "    outf.write('#' + '\\t'.join(('SampleID', 'Metaphlan_Analysis')) + '\\n')\n",
    "    outf.write('#clade_name\\tNCBI_tax_id\\trelative_abundance\\tadditional_species\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unsupported operand type(s) for *: 'NoneType' and 'int'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[29], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m cl2ab, _, tot_nreads \u001b[39m=\u001b[39m tree\u001b[39m.\u001b[39;49mrelative_abundances(\n\u001b[1;32m      2\u001b[0m                         \u001b[39m'\u001b[39;49m\u001b[39ma\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39m+\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m__\u001b[39;49m\u001b[39m\"\u001b[39;49m \u001b[39mif\u001b[39;49;00m \u001b[39m'\u001b[39;49m\u001b[39mtax_lev\u001b[39;49m\u001b[39m'\u001b[39;49m \u001b[39m!=\u001b[39;49m \u001b[39m'\u001b[39;49m\u001b[39ma\u001b[39;49m\u001b[39m'\u001b[39;49m \u001b[39melse\u001b[39;49;00m \u001b[39mNone\u001b[39;49;00m )\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:767\u001b[0m, in \u001b[0;36mTaxTree.relative_abundances\u001b[0;34m(self, tax_lev)\u001b[0m\n\u001b[1;32m    764\u001b[0m clade2abundance, clade2est_nreads, tot_ab, tot_reads \u001b[39m=\u001b[39m {}, {}, \u001b[39m0.0\u001b[39m, \u001b[39m0\u001b[39m\n\u001b[1;32m    766\u001b[0m \u001b[39mfor\u001b[39;00m tax_label, clade \u001b[39min\u001b[39;00m clade2abundance_n\u001b[39m.\u001b[39mitems():\n\u001b[0;32m--> 767\u001b[0m     tot_ab \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m clade\u001b[39m.\u001b[39;49mcompute_abundance()\n\u001b[1;32m    769\u001b[0m \u001b[39mfor\u001b[39;00m tax_label, clade \u001b[39min\u001b[39;00m clade2abundance_n\u001b[39m.\u001b[39mitems():\n\u001b[1;32m    770\u001b[0m     \u001b[39mfor\u001b[39;00m clade_label, tax_id, abundance \u001b[39min\u001b[39;00m \u001b[39msorted\u001b[39m(clade\u001b[39m.\u001b[39mget_all_abundances(), key\u001b[39m=\u001b[39m\u001b[39mlambda\u001b[39;00m pars:pars[\u001b[39m0\u001b[39m]):\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:538\u001b[0m, in \u001b[0;36mTaxClade.compute_abundance\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcompute_abundance\u001b[39m( \u001b[39mself\u001b[39m ):\n\u001b[1;32m    536\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m: \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance\n\u001b[0;32m--> 538\u001b[0m     sum_ab \u001b[39m=\u001b[39m \u001b[39msum\u001b[39m([c\u001b[39m.\u001b[39mcompute_abundance() \u001b[39mfor\u001b[39;00m c \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchildren\u001b[39m.\u001b[39mvalues()])\n\u001b[1;32m    540\u001b[0m     \u001b[39m# rat_nreads = sorted([(self.markers2lens[marker], n_reads)\u001b[39;00m\n\u001b[1;32m    541\u001b[0m     \u001b[39m#                             for marker,n_reads in self.markers2nreads.items()],\u001b[39;00m\n\u001b[1;32m    542\u001b[0m     \u001b[39m#                                     key = lambda x: x[1])\u001b[39;00m\n\u001b[1;32m    544\u001b[0m     rat_nreads, removed \u001b[39m=\u001b[39m [], []\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:538\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcompute_abundance\u001b[39m( \u001b[39mself\u001b[39m ):\n\u001b[1;32m    536\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m: \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance\n\u001b[0;32m--> 538\u001b[0m     sum_ab \u001b[39m=\u001b[39m \u001b[39msum\u001b[39m([c\u001b[39m.\u001b[39;49mcompute_abundance() \u001b[39mfor\u001b[39;00m c \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchildren\u001b[39m.\u001b[39mvalues()])\n\u001b[1;32m    540\u001b[0m     \u001b[39m# rat_nreads = sorted([(self.markers2lens[marker], n_reads)\u001b[39;00m\n\u001b[1;32m    541\u001b[0m     \u001b[39m#                             for marker,n_reads in self.markers2nreads.items()],\u001b[39;00m\n\u001b[1;32m    542\u001b[0m     \u001b[39m#                                     key = lambda x: x[1])\u001b[39;00m\n\u001b[1;32m    544\u001b[0m     rat_nreads, removed \u001b[39m=\u001b[39m [], []\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:538\u001b[0m, in \u001b[0;36mTaxClade.compute_abundance\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcompute_abundance\u001b[39m( \u001b[39mself\u001b[39m ):\n\u001b[1;32m    536\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m: \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance\n\u001b[0;32m--> 538\u001b[0m     sum_ab \u001b[39m=\u001b[39m \u001b[39msum\u001b[39m([c\u001b[39m.\u001b[39mcompute_abundance() \u001b[39mfor\u001b[39;00m c \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchildren\u001b[39m.\u001b[39mvalues()])\n\u001b[1;32m    540\u001b[0m     \u001b[39m# rat_nreads = sorted([(self.markers2lens[marker], n_reads)\u001b[39;00m\n\u001b[1;32m    541\u001b[0m     \u001b[39m#                             for marker,n_reads in self.markers2nreads.items()],\u001b[39;00m\n\u001b[1;32m    542\u001b[0m     \u001b[39m#                                     key = lambda x: x[1])\u001b[39;00m\n\u001b[1;32m    544\u001b[0m     rat_nreads, removed \u001b[39m=\u001b[39m [], []\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:538\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcompute_abundance\u001b[39m( \u001b[39mself\u001b[39m ):\n\u001b[1;32m    536\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m: \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance\n\u001b[0;32m--> 538\u001b[0m     sum_ab \u001b[39m=\u001b[39m \u001b[39msum\u001b[39m([c\u001b[39m.\u001b[39;49mcompute_abundance() \u001b[39mfor\u001b[39;00m c \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchildren\u001b[39m.\u001b[39mvalues()])\n\u001b[1;32m    540\u001b[0m     \u001b[39m# rat_nreads = sorted([(self.markers2lens[marker], n_reads)\u001b[39;00m\n\u001b[1;32m    541\u001b[0m     \u001b[39m#                             for marker,n_reads in self.markers2nreads.items()],\u001b[39;00m\n\u001b[1;32m    542\u001b[0m     \u001b[39m#                                     key = lambda x: x[1])\u001b[39;00m\n\u001b[1;32m    544\u001b[0m     rat_nreads, removed \u001b[39m=\u001b[39m [], []\n",
      "    \u001b[0;31m[... skipping similar frames: <listcomp> at line 538 (4 times), TaxClade.compute_abundance at line 538 (4 times)]\u001b[0m\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:538\u001b[0m, in \u001b[0;36mTaxClade.compute_abundance\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcompute_abundance\u001b[39m( \u001b[39mself\u001b[39m ):\n\u001b[1;32m    536\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m: \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance\n\u001b[0;32m--> 538\u001b[0m     sum_ab \u001b[39m=\u001b[39m \u001b[39msum\u001b[39m([c\u001b[39m.\u001b[39mcompute_abundance() \u001b[39mfor\u001b[39;00m c \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchildren\u001b[39m.\u001b[39mvalues()])\n\u001b[1;32m    540\u001b[0m     \u001b[39m# rat_nreads = sorted([(self.markers2lens[marker], n_reads)\u001b[39;00m\n\u001b[1;32m    541\u001b[0m     \u001b[39m#                             for marker,n_reads in self.markers2nreads.items()],\u001b[39;00m\n\u001b[1;32m    542\u001b[0m     \u001b[39m#                                     key = lambda x: x[1])\u001b[39;00m\n\u001b[1;32m    544\u001b[0m     rat_nreads, removed \u001b[39m=\u001b[39m [], []\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:538\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcompute_abundance\u001b[39m( \u001b[39mself\u001b[39m ):\n\u001b[1;32m    536\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m: \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mabundance\n\u001b[0;32m--> 538\u001b[0m     sum_ab \u001b[39m=\u001b[39m \u001b[39msum\u001b[39m([c\u001b[39m.\u001b[39;49mcompute_abundance() \u001b[39mfor\u001b[39;00m c \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchildren\u001b[39m.\u001b[39mvalues()])\n\u001b[1;32m    540\u001b[0m     \u001b[39m# rat_nreads = sorted([(self.markers2lens[marker], n_reads)\u001b[39;00m\n\u001b[1;32m    541\u001b[0m     \u001b[39m#                             for marker,n_reads in self.markers2nreads.items()],\u001b[39;00m\n\u001b[1;32m    542\u001b[0m     \u001b[39m#                                     key = lambda x: x[1])\u001b[39;00m\n\u001b[1;32m    544\u001b[0m     rat_nreads, removed \u001b[39m=\u001b[39m [], []\n",
      "File \u001b[0;32m/data/wangyang/MetaPhlAn/metaphlan/metaphlan.py:587\u001b[0m, in \u001b[0;36mTaxClade.compute_abundance\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    585\u001b[0m rat_v,nreads_v \u001b[39m=\u001b[39m \u001b[39mzip\u001b[39m(\u001b[39m*\u001b[39mrat_nreads) \u001b[39mif\u001b[39;00m rat_nreads \u001b[39melse\u001b[39;00m ([],[])\n\u001b[1;32m    586\u001b[0m rat, nrawreads, loc_ab \u001b[39m=\u001b[39m \u001b[39mfloat\u001b[39m(\u001b[39msum\u001b[39m(rat_v)) \u001b[39mor\u001b[39;00m \u001b[39m-\u001b[39m\u001b[39m1.0\u001b[39m, \u001b[39msum\u001b[39m(nreads_v), \u001b[39m0.0\u001b[39m\n\u001b[0;32m--> 587\u001b[0m quant \u001b[39m=\u001b[39m \u001b[39mint\u001b[39m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mquantile\u001b[39m*\u001b[39;49m\u001b[39mlen\u001b[39;49m(rat_nreads))\n\u001b[1;32m    588\u001b[0m ql,qr,qn \u001b[39m=\u001b[39m (quant,\u001b[39m-\u001b[39mquant,quant) \u001b[39mif\u001b[39;00m quant \u001b[39melse\u001b[39;00m (\u001b[39mNone\u001b[39;00m,\u001b[39mNone\u001b[39;00m,\u001b[39m0\u001b[39m)\n\u001b[1;32m    590\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m SGB_ANALYSIS \u001b[39mand\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mname[\u001b[39m0\u001b[39m] \u001b[39m==\u001b[39m \u001b[39m'\u001b[39m\u001b[39mt\u001b[39m\u001b[39m'\u001b[39m \u001b[39mand\u001b[39;00m (\u001b[39mlen\u001b[39m(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfather\u001b[39m.\u001b[39mchildren) \u001b[39m>\u001b[39m \u001b[39m1\u001b[39m \u001b[39mor\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39m_sp\u001b[39m\u001b[39m\"\u001b[39m \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfather\u001b[39m.\u001b[39mname \u001b[39mor\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39mk__Viruses\u001b[39m\u001b[39m\"\u001b[39m \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mget_full_name()):\n",
      "\u001b[0;31mTypeError\u001b[0m: unsupported operand type(s) for *: 'NoneType' and 'int'"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'a__'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'a'+\"__\" if 'tax_lev' != 'a' else None "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "metaphlan",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "907b62017a02795733071a30e34ccf801a36defa4c7985cf6a3a3f7e520a1fb5"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
